sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+    typename std::enable_if<__check_atomic<Op, T, n, 0>(), simd<T, n>>::type
+    flat_atomic(T *p, simd<unsigned, n> offset, simd<ushort, n> pred);
+```
+
+- ```Flat-address atomic add/sub/min/max/etc.```: perform atomic memory
+access operation with one source operand.
+
+```cpp
+template <CmAtomicOpType Op, typename T, int n,
+          CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+    typename std::enable_if<__check_atomic<Op, T, n, 1>(), simd<T, n>>::type
+    flat_atomic(T *p, simd<unsigned, n> offset, simd<T, n> src0,
+                simd<ushort, n> pred);
+```
+
+- ```Flat-address atomic CMPXCHG```: perform atomic memory access operation
+with two source operands.
+
+```cpp
+
+template <CmAtomicOpType Op, typename T, int n,
+          CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+    typename std::enable_if<__check_atomic<Op, T, n, 2>(), simd<T, n>>::type
+    flat_atomic(T *p, simd<unsigned, n> offset, simd<T, n> src0,
+                simd<T, n> src1, simd<ushort, n> pred);
+```
+
+#### Accessor-based memory access
+
+Examples:
+##### 1D surface block store.
+T � element type, n � vector length, acc � SYCL global buffer accessor,
+offset � byte offset in the buffer, vals � vector value to store
+```cpp
+template <typename T, int n, typename AccessorTy>
+void block_store(AccessorTy acc, uint32_t offset, simd<T, n> vals);
+```
+##### 2D media block load.
+T � element type, m and n � block dimensions, acc � SYCL image2D accessor,
+x and y � image coordinates
+```cpp
+template <typename T, int m, int n, typename AccessorTy, unsigned plane = 0>
+simd<T, m * n> media_block_load(AccessorTy acc, unsigned x, unsigned y);
+```
+
+#### Local address space allocation and access
+Examples:
+
+##### SLM scatter.
+T � element type, n � vector length (16 or 32), vals � vector value to store,
+offsets � byte offsets in the local memory, pred � store mask
+```cpp
+template <typename T, int n>
+typename std::enable_if<(n == 16 || n == 32), void>::type
+slm_store(simd<T, n> vals, simd<uint32_t, n> offsets, simd<uint16_t, n> pred = 1);
+```
+<br>
+
+### Private Global Variables.
+
+Explicit SIMD extenstion supports "private global" variables - file scope
+variables in private address space (similar to thread-local variables on host).
+These variables have 1 copy per work-item (which maps to a single SIMD thread in
+ESP) and are visible to all functions in the translation unit. Conceptually they
+map to SPIRV variable with private storage class. Private globals can be bound
+to a specific byte offset within the GRF. To mark a file scope variable as
+private global, the `INTEL_GPU_PRIVATE` attribute macro is used,
+`INTEL_GPU_REGISTER` is used to bind it the register file:
+
+```cpp
+INTEL_GPU_PRIVATE INTEL_GPU_REGISTER(32) simd<int, 16> vc;
+```
+<br>
+
+## Examples
+### Vector addition (USM)
+```cpp
+#include <iostream>
+#include <CL/sycl.hpp>
+#include <sycl_explicit_simd.hpp>
```

Comment:
Why is this not included automatically when the extension is enabled?

Reply:
Following mainstream SYCL - sycl.hpp is not auto included under -fsycl. I can add a TODO and file an issue.

Reply:
That's a good argument, but we currently do automatically include things like the sub-group and USM headers.  I don't mind which way we go as long as we're consistent; marking this with a TODO and filing an issue to address this later works for me.

Reply:
Filed internal (for now) issue, added a TODO